elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/ElasticsearchSourceStage.scala

```diff
@@ -0,0 +1,148 @@
+
+  val out: Outlet[OutgoingMessage[JsObject]] = Outlet("ElasticsearchSource.out")
+  override val shape: SourceShape[OutgoingMessage[JsObject]] = SourceShape(out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new ElasticsearchSourceLogic[JsObject](indexName, typeName, query, client, settings, out, shape) {
+      override protected def convert(jsObj: JsObject): JsObject = jsObj
+    }
+
+}
+
+final class ElasticsearchSourceStageTyped[T](indexName: String,
+                                             typeName: String,
+                                             query: String,
+                                             client: RestClient,
+                                             settings: ElasticsearchSourceSettings)(implicit reader: JsonReader[T])
+    extends GraphStage[SourceShape[OutgoingMessage[T]]] {
+
+  val out: Outlet[OutgoingMessage[T]] = Outlet("ElasticsearchSource.out")
+  override val shape: SourceShape[OutgoingMessage[T]] = SourceShape(out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new ElasticsearchSourceLogic[T](indexName, typeName, query, client, settings, out, shape) {
+      override protected def convert(jsObj: JsObject): T = jsObj.convertTo[T]
+    }
+
+}
+
+sealed abstract class ElasticsearchSourceLogic[T](indexName: String,
+                                                  typeName: String,
+                                                  query: String,
+                                                  client: RestClient,
+                                                  settings: ElasticsearchSourceSettings,
+                                                  out: Outlet[OutgoingMessage[T]],
+                                                  shape: SourceShape[OutgoingMessage[T]])
+    extends GraphStageLogic(shape) {
+
+  private var started = false
+  private var scrollId: String = null
+  private val responseHandler = getAsyncCallback[Response](handleResponse)
+  private val failureHandler = getAsyncCallback[Throwable](handleFailure)
+
+  protected def convert(jsObj: JsObject): T
+
+  def sendScrollScanRequest(): Unit =
+    try {
+      if (scrollId == null) {
+        client.performRequestAsync(
+          "POST",
+          s"$indexName/$typeName/_search",
+          Map("scroll" -> "5m", "sort" -> "_doc").asJava,
+          new StringEntity(s"""{"size": ${settings.bufferSize}, "query": ${query}}"""),
+          new ResponseListener {
+            override def onFailure(exception: Exception) = failureHandler.invoke(exception)
+            override def onSuccess(response: Response) = responseHandler.invoke(response)
+          }
+        )
+      } else {
+        client.performRequestAsync(
+          "POST",
+          s"/_search/scroll",
+          Map[String, String]().asJava,
+          new StringEntity(Map("scroll" -> "5m", "scroll_id" -> scrollId).toJson.toString),
+          new ResponseListener {
```

Comment:
The GraphStageLogic could implement the `ResponseListener` directly. That would also remove the code duplication with the above branch.